<a href="https://colab.research.google.com/github/moeuu/colab/blob/main/algorithm/2/sqrt_bisection_method.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 平方根の見積もりのための2分法

## 平方根の見積もりのための2分法

[平方根の見積もりに対する2分法](https://drive.google.com/file/d/1rAwJPMT1gKC3E2AHdQFeMh7o_YbKcxWE/view?usp=sharing)を関数square_root_bisection_methodとして以下に定義する．

In [ ]:
def square_root_bisection_method(x, p):
    a, b = 0, x
    while b - a > p:
        m = (a + b) / 2 # このコードはPython3用なので，この商は小数部分切り捨てではない．
        if m ** 2 > x:
            b = m
        else:
            a = m
    return a, b

試しに，$\sqrt{2}$を精度0.0001で見積もってみる．

In [ ]:
square_root_bisection_method(2, 0.0001)

(1.4141845703125, 1.41424560546875)

それらしい値が出力された．

アルゴリズムの途中経過を観察するための表示命令を加えた関数をchatty_square_root_bisection_methodとして以下に定義する．

In [ ]:
def chatty_square_root_bisection_method(x, p):
    a, b = 0, x
    while b - a > p:
        print(f'[a, b] = [{a:4.3f}, {b:4.3f}]') # この命令でアルゴリズムの途中経過を観察する．
        # {a:4.3f}は「変数aの値を，少数(f)で，全部で4桁で，小数点以下は3桁で揃えて表示」という意味である．
        m = (a + b) / 2 # このコードはPython3用なので，この商は小数部分切り捨てではない．
        if m ** 2 > x:
            b = m
        else:
            a = m
    return a, b

試しに，$\sqrt{2}$を精度0.1で見積もってみる．

In [ ]:
chatty_square_root_bisection_method(2, 0.1)

[a, b] = [0.000, 2.000]
[a, b] = [1.000, 2.000]
[a, b] = [1.000, 1.500]
[a, b] = [1.250, 1.500]
[a, b] = [1.375, 1.500]


(1.375, 1.4375)

区間$[a, b]$に$\sqrt{2}$を含みながら，$b - a$が半分ずつになる様子が見て取れる．

そして見積もりとしては甘く感じるが，$\sqrt{2}$は1.375と1.4375の間にあるとわかる．

もう少し精度を上げて，$\sqrt{2}$を精度0.0001で見積もってみる．

In [ ]:
chatty_square_root_bisection_method(2, 0.0001)

[a, b] = [0.000, 2.000]
[a, b] = [1.000, 2.000]
[a, b] = [1.000, 1.500]
[a, b] = [1.250, 1.500]
[a, b] = [1.375, 1.500]
[a, b] = [1.375, 1.438]
[a, b] = [1.406, 1.438]
[a, b] = [1.406, 1.422]
[a, b] = [1.414, 1.422]
[a, b] = [1.414, 1.418]
[a, b] = [1.414, 1.416]
[a, b] = [1.414, 1.415]
[a, b] = [1.414, 1.415]
[a, b] = [1.414, 1.414]
[a, b] = [1.414, 1.414]


(1.4141845703125, 1.41424560546875)

2分法を用いると，目標を含む区間が毎回半分になる．
よって探索範囲が狭まるのは意外と早い．

## 再帰アルゴリズムとしての2分法

再帰アルゴリズムとしての平方根の見積もりのための2分法を関数recursive_square_root_bisection_methodとして以下に定義する．
関数recursive_square_root_bisection_methodは再帰関数recursive_square_root_bisectionを呼び出すだけの関数なので，実質的な再帰アルゴリズムを表しているのは再帰関数recursive_square_root_bisectionである．

In [ ]:
def recursive_square_root_bisection_method(x, p): # この関数はサブルーチンとしての再帰関数を呼び出すだけ
    # このように実質的な働きはなく，引数のつじつまを合わせるだけの関数をラッパー関数などと呼んだりする．
    return recursive_square_root_bisection(x, p, 0, x)

def recursive_square_root_bisection(x, p, a, b): # この関数が実質的な再帰アルゴリズム
    if b - a <= p:
        return a, b
    m = (a + b) / 2 # このコードはPython3用なので，この商は小数部分切り捨てではない．
    if m ** 2 > x:
        return recursive_square_root_bisection(x, p, a, m)
    else:
        return recursive_square_root_bisection(x, p, m, b)

In [ ]:
recursive_square_root_bisection_method(2, 0.0001)

(1.4141845703125, 1.41424560546875)